airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionReader.kt

```diff
@@ -0,0 +1,203 @@
+                        // In that case, we still want to fail loud and clear
+                        log.info { "Debezium failed with: $message" }
+                    }
+                }
+            }
+            .using(
+                object : DebeziumEngine.ConnectorCallback {
+                    override fun connectorStarted() {
+                        log.info { "DebeziumEngine notify: connector started" }
+                    }
+
+                    override fun connectorStopped() {
+                        log.info { "DebeziumEngine notify: connector stopped" }
+                    }
+
+                    override fun taskStarted() {
+                        log.info { "DebeziumEngine notify: task started" }
+                    }
+
+                    override fun taskStopped() {
+                        log.info { "DebeziumEngine notify: task stopped" }
+                    }
+                },
+            )
+            .build()
+    }
+
+    /**
+     * [DebeziumRecordIterator.heartbeatEventSourceField] acts as a cache so that we avoid using
+     * reflection to setAccessible for each event
+     */
+    @VisibleForTesting
+    internal fun getSourceRecord(heartbeatEvent: ChangeEvent<String?, String?>): SourceRecord {
+        try {
+            val eventClass: Class<out ChangeEvent<*, *>?> = heartbeatEvent.javaClass
+            val f: Field?
+            if (heartbeatEventSourceField.containsKey(eventClass)) {
+                f = heartbeatEventSourceField[eventClass]
+            } else {
+                f = eventClass.getDeclaredField("sourceRecord")
+                f.isAccessible = true
+                heartbeatEventSourceField[eventClass] = f
+
+                if (heartbeatEventSourceField.size > 1) {
+                    log.warn {
+                        "Field Cache size growing beyond expected size of 1, size is ${heartbeatEventSourceField.size}"
+                    }
+                }
+            }
+
+            val sr = f!![heartbeatEvent] as SourceRecord
+            return sr
+        } catch (e: NoSuchFieldException) {
+            log.info { "failed to get heartbeat source offset" }
+            throw RuntimeException(e)
+        } catch (e: IllegalAccessException) {
+            log.info { "failed to get heartbeat source offset" }
+            throw RuntimeException(e)
+        }
+    }
+
+    enum class DebeziumCloseReason() {
+        TIMEOUT,
+        HEARTBEAT_REACHED_TARGET_POSITION,
+        CHANGE_EVENT_REACHED_TARGET_POSITION,
+        HEARTBEAT_NOT_PROGRESSING
+    }
+
+    private fun requestClose(closeLogMessage: String, closeReason: DebeziumCloseReason) {
+        log.info { closeLogMessage }
+        log.info { closeReason }
+        // TODO : send close analytics message
+        // TODO : Close the engine. Note that engine must be closed in a different thread
+        // than the running engine.
+        CoroutineScope(Dispatchers.IO).launch { engine?.close() }
```

Comment:
Yes. This is how I understand this as well.
Because debezium never `suspend`s it will remain on its thread for lifetime.
So `launch`ing something is guaranteed to run on a different thread.